x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SuccessfulAuthenticationResponseMessageBuilder.java

```diff
@@ -0,0 +1,219 @@
+        if (attributes != null) {
+            assertion.getAttributeStatements().add(attributes);
+        }
+        response.getAssertions().add(assertion);
+
+        return sign(response);
+    }
+
+    private Response sign(Response response) {
+        // TODO
+        return response;
+    }
+
+    private Conditions buildConditions(DateTime now, SamlServiceProvider serviceProvider) {
+        final Audience spAudience = samlFactory.object(Audience.class, Audience.DEFAULT_ELEMENT_NAME);
+        spAudience.setAudienceURI(serviceProvider.getEntityId());
+
+        final AudienceRestriction restriction = samlFactory.object(AudienceRestriction.class, AudienceRestriction.DEFAULT_ELEMENT_NAME);
+        restriction.getAudiences().add(spAudience);
+
+        final Conditions conditions = samlFactory.object(Conditions.class, Conditions.DEFAULT_ELEMENT_NAME);
+        conditions.setNotBefore(now);
+        conditions.setNotOnOrAfter(now.plus(serviceProvider.getAuthnExpiry()));
+        conditions.getAudienceRestrictions().add(restriction);
+        return conditions;
+    }
+
+    private DateTime now() {
+        return new DateTime(clock.millis(), DateTimeZone.UTC);
+    }
+
+    private Subject buildSubject(DateTime now, UserServiceAuthentication user, AuthnRequest request) {
+        final SamlServiceProvider serviceProvider = user.getServiceProvider();
+
+        final NameID nameID = samlFactory.object(NameID.class, NameID.DEFAULT_ELEMENT_NAME);
+        nameID.setFormat(NameIDType.PERSISTENT);
+        nameID.setValue(user.getPrincipal());
+
+        final Subject subject = samlFactory.object(Subject.class, Subject.DEFAULT_ELEMENT_NAME);
+        subject.setNameID(nameID);
+
+        final SubjectConfirmationData data = samlFactory.object(SubjectConfirmationData.class,
+            SubjectConfirmationData.DEFAULT_ELEMENT_NAME);
+        if (request != null) {
+            data.setInResponseTo(request.getID());
+        }
+        data.setNotBefore(now);
+        data.setNotOnOrAfter(now.plus(serviceProvider.getAuthnExpiry()));
+        data.setRecipient(serviceProvider.getAssertionConsumerService().toString());
+
+        final SubjectConfirmation confirmation = samlFactory.object(SubjectConfirmation.class, SubjectConfirmation.DEFAULT_ELEMENT_NAME);
+        confirmation.setMethod(SubjectConfirmation.METHOD_BEARER);
+        confirmation.setSubjectConfirmationData(data);
+
+        subject.getSubjectConfirmations().add(confirmation);
+        return subject;
+    }
+
+    private AuthnStatement buildAuthnStatement(DateTime now, UserServiceAuthentication user) {
+        final SamlServiceProvider serviceProvider = user.getServiceProvider();
+        final AuthnStatement statement = samlFactory.object(AuthnStatement.class, AuthnStatement.DEFAULT_ELEMENT_NAME);
+        statement.setAuthnInstant(now);
+        statement.setSessionNotOnOrAfter(now.plus(serviceProvider.getAuthnExpiry()));
```

Comment:
Sorry, it was late when I finished up so I didn't comment.
I agree we'll need one for SLO, but since we're not doing SLO yet, I intentionally left it out so that whoever implements SLO can do it end-to-end.
That's on the assumption that there's nothing in the standard that requires a `SessionIndex` (I didn't see anything).

Reply:
+1